### DISCUSSION TITLE
Using Parcel 2 programatically

### DISCUSSION THREAD

**Discussion by m-allanson** [parcel-bundler/parcel#4804] [General] [answered]
I am trying to use `@parcel/core` to programatically create some bundles in an existing project. See here for an example config: https://github.com/m-allanson/parcelish/blob/master/scripts/bundle.js

However I'm running into an issue where the project's `package.json` file seems to be overriding some of the options passed into Parcel. 

Here's roughly how I'm running Parcel:

```js
// bundle.js
let bundler = new Parcel({
  defaultConfig: {
    ...defaultConfigContents,
    filePath: require.resolve("@parcel/config-default"),
  },
  entries: "src/target.js",
  isLibrary: true,
  mode: "production",
  outputFormat: "esmodule",
  target: {
    outputFormat: "esmodule",
    distDir: "dist",
  },
});

await bundler.run();
```

Note that the `package.json` contains a `main` entry, but I don't want Parcel to use it. 

```js
{
  "name": "example-project",
  "main": "src/index.js",
// ...
```

When I run the bundling, Parcel writes each bundle to `src/index.js`, instead of writing to the `dist` directory. 

If I remove `"main": "src/index.js",` from the `package.json`, Parcel will output to the `dist` directory instead. I'd like this behaviour without having to remove the `main` property from the `package.json` file.

My questions:

- Is this expected behaviour, or am I doing something odd?
- Can I tell Parcel to ignore the `package.json` file, and only use the options that are provided by `bundle.js`?
- Can I manually specify a destination for a target? The API docs mention a [`distEntry` value for Target](https://v2.parceljs.org/plugin-system/api/#Target). However, that value is [not present on PackageTargetDescriptor](https://v2.parceljs.org/plugin-system/api/#PackageTargetDescriptor), so I can't see a way of manually specifying it. 

Here's a runnable example: https://github.com/m-allanson/parcelish

**Comment by mischnic** [comment, marked as answer]
> Can I tell Parcel to ignore the package.json file, and only use the options that are provided by bundle.js?

I've looked at the code and passing a targets property in the options should cause package.json to be ignored. Not sure why this happens

> However, that value is not present on PackageTargetDescriptor,

https://v2.parceljs.org/plugin-system/api/#PackageTargetDescriptor is used in a package.json, where `package.json#default` would be used as the distEntry for `package.json#targets.default`

So I guess it should be an optional property here: https://v2.parceljs.org/plugin-system/api/#TargetDescriptor

**Reply by m-allanson** [reply]
Thanks! I’ll investigate some more later next week.

**Comment by navaru** [comment]
Change
```js
for (const module in config) {
  let task = async () => {
    let bundler = new Parcel({
      entries: config[module].entry,
      target: config[module].target, // <-- here is the problem
```
to

```js
for (const module in config) {
  let task = async () => {
    let bundler = new Parcel({
      entries: config[module].entry,
      targets: { [module]: config[module].target },  // <-- this is the fix
```

**Reply by m-allanson** [reply]
😱 thank you @navaru! That did the trick.

**Comment by m-allanson** [comment]
Thanks both for your replies. For anyone looking to use Parcel 2 like this, here's the working code that I ended up with:

https://github.com/stylelint/stylelint/blob/642bd3201277a42e9cf8cd705a663795d92b32bd/scripts/bundle-esm.js

Note that Parcel [has a private API `_getResolvedParcelOptions`](https://github.com/parcel-bundler/parcel/blob/78b7910149440f34faa7eb636263e98967a8eb22/packages/core/core/src/Parcel.js#L380-L387) that you can use to verify that your options are resolved as you expected. It can be useful to help you understand why Parcel is not doing what you want it to :)

In my case I was not specifying the targets correctly, so I think Parcel was falling back to checking the `package.json` file for that info.

**Comment by ivoba** [comment]
I have a similar problem, with also a main entry in package.json that should be ignored.

My setup:

     "@parcel/config-default": "^2.0.0-nightly.778",
     "@parcel/core": "^2.0.0-nightly.776",

calling Parcel via API:

```
const parcel = new Parcel({
    entries: path.join(__dirname, "../views/index.html"),
    defaultConfig: require.resolve("@parcel/config-default"),
    logLevel: 'verbose',
    targets: {
      default: {
        distDir: path.join(__dirname, "../dist"),
        distEntry: 'index.html',
        source: path.join(__dirname, "../views/index.html")
      },
    },
    mode: "production",
  })

await parcel.run().catch(err => console.log(err))
```

But there is no output at all, neither in console nor are files generated.

Also it seems redundant to declare entries and again in targets.
I am also not sure which target is called eventually?

```
node -v
v14.16.1
```

Any Ideas?

**Comment by padcom** [comment]
@ivoba Are you by any chance trying to get it to work with Electron app? That Node.js version 14.16 looks familiar :)
